[REEF-1312] Convert IMRU.Examples to test

tcNickolas opened the pull request:
This change creates tests which call IMRU.Examples and verify the run succeeded.
At the moment only local runtime tests are operational, as none of our
functional tests can run on Yarn yet.

JIRA:
  [REEF-1312](https://issues.apache.org/jira/browse/REEF-1312)

Pull request:
  This closes #

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceTest.cs:
```diff
@@ -0,0 +1,73 @@
+        [Trait("Description", "Run IMRU broadcast and reduce example as test on Yarn.")]
+        void TestIMRUBroadcastReduceOnYarn()
+        {
+            string testFolder = DefaultRuntimeFolder + TestId + "Yarn";
```
It is not used in Yarn case.

tcNickolas replied:
Looks like it's not used in other OnYarn tests also. Should we file a JIRA to clean up all OnYarn tests to improve clarity?

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceTest.cs:
```diff
@@ -0,0 +1,73 @@
+            TestIMRUBroadcastReduce(true, testFolder);
+        }
+
+        private void TestIMRUBroadcastReduce(bool runOnYarn, string testFolder)
```
You can also use params string[]  for test folder here so that you don't need to pass to it for Yarn case.

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUMapperCountTest.cs:
```diff
@@ -0,0 +1,59 @@
+            TestIMRUMapperCount(true, testFolder);
+        }
+
+        private void TestIMRUMapperCount(bool runOnYarn, string testFolder)
```
Same, you might change the last parameter as params string[] so that not to pass it for Yarn case as it is not used for Yarn and may cause confusion if test passes it.

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Org.Apache.REEF.Tests.csproj:
```diff
@@ -157,6 +159,10 @@ under the License.
       <Project>{159f7d70-8acc-4d97-9f6d-2fc4ca0d8682}</Project>
       <Name>Org.Apache.REEF.Examples.AllHandlers</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Org.Apache.REEF.IMRU.Examples\Org.Apache.REEF.IMRU.Examples.csproj">
```
Use $(SolutionDir) to replace ..

jwang98052 commented:
I complete a pass. 
@dkm2110 please review it.

dkm2110 commented:
@tcNickolas I had an overall look. One comment: There are additional useful logs in BroadcastReduce examples that you can test for. Specifically, in Org.Apache.REEF.IMRU.Examples.PipelinedBroadcastReduce.BroadcastSenderReduceReceiverUpdateFunction
we print the value received in each iteration. Might be worthwhile to check that.

Similarly in Org.Apache.REEF.IMRU.Examples.MapperCount.MapperCountUpdateFunction, the input in Update function there is basically total mapper count. We can introduce a log in that function and then check the value.

tcNickolas commented:
I've addressed the comments and pushed rebased version

@dkm2110 I'm reluctant to read multiple evaluator logs to verify inputs of functions. First, tests should verify outputs, not inputs, and ideally as Assert in the test task itself instead of logs. Second, our current infrastructure supports reading at most 1 log for verification, and I'd rather invest effort into new infrastructure that would not be log-based (we have a JIRA for that) than in adding more log reading. Let's keep this as is for now

dkm2110 commented:
I see OK. I do not have objection then.

tcNickolas commented:
@dkm2110 @jwang98052 Can you please review this and merge if there are no other comments?

markusweimer commented:
I'll test and merge. I want these tests in to avoid the kind of situation we currently discuss on the list :)

tcNickolas commented:
Yes, it is my wish also :-) Thank you!

markusweimer commented:
It might take a while, as I want to run the tests several times. Just to be sure.

markusweimer commented:
I had one test failure in 4 runs. The html report contained this summary:

```
Org.Apache.REEF.Tests.Functional.IMRU.IMRUMapperCountTest.TestIMRUMapperCountOnLocalRuntime

Expected number of contexts to close (4) differs from actual number of success indicators (8)\r\nExpected: True\r\nActual:   False

   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.ValidateSuccessForLocalRuntime(Int32 numberOfContextsToClose, Int32 numberOfTasksToFail, Int32 numberOfEvaluatorsToFail, String testFolder) in D:\src\reef\lang\cs\Org.Apache.REEF.Tests\Functional\ReefFunctionalTest.cs:line 179
   at Org.Apache.REEF.Tests.Functional.IMRU.IMRUMapperCountTest.TestIMRUMapperCountOnLocalRuntime() in D:\src\reef\lang\cs\Org.Apache.REEF.Tests\Functional\IMRU\IMRUMapperCountTest.cs:line 38
```

markusweimer commented:
One more data point: I did not recompile the project between runs. This means that the non-determinism isn't in the build process, but the tests themselves.

tcNickolas commented:
Interesting.
@dkm2110 You know this code best. Is it possible that sometimes IMRU opens two contexts per evaluator? Or requests more evaluators than configured?

markusweimer commented:
Update on my test runs: I had 1 failure in 8 runs so far.

dkm2110 commented:
Mariia

Nothing like this should happen. However, I fear there might be some stream
closing issues leading to failures. REEF-1224 should address that.

Dhruv

On Wed, Apr 13, 2016 at 10:37 AM, Mariia Mykhailova <
notifications@github.com> wrote:

> Interesting.
> @dkm2110 <https://github.com/dkm2110> You know this code best. Is it
> possible that sometimes IMRU opens two contexts per evaluator? Or requests
> more evaluators than configured?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/940#issuecomment-209558431>
>

dkm2110 commented:
If you could send me the failed runtime dir I can throw some light.

On Wed, Apr 13, 2016 at 10:58 AM, Markus Weimer <notifications@github.com>
wrote:

> Update on my test runs: I had 1 failure in 8 runs so far.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/940#issuecomment-209569493>
>

markusweimer commented:
I now have 1 out of 12 tests that failed. I'll stop running tests, as this seems to be very rare.

markusweimer commented:
@dkm2110 I uploaded it to [OneDrive](https://onedrive.live.com/redir?resid=5801726772BFC3DA!273888&authkey=!ACMrVHlIHAHvCi8&ithint=file%2czip). Just be aware that it is a 160MB file.

dkm2110 commented:
Seems like a reef bug. look in node 1

Org.Apache.REEF.Common.Runtime.Evaluator.EvaluatorRuntime Error: 0 :
2016-04-13T10:24:03.5192372-07:00 0007
ERROR: evaluator Node-1-1460568240750 failed with exceptionencountered
error [System.InvalidOperationException: Received a control message from
Driver after Evaluator is done.] with mesage [Received a control message
from Driver after Evaluator is done.] and stack trace []
Org.Apache.REEF.Common.Runtime.Evaluator.HeartBeatManager Information: 0 :
2016-04-13T10:24:03.5197049-07:00 0007
INFO: Triggered a heartbeat: EvaluatorHeartbeatProto: task_id=[],
task_status=[], task_message=[], evaluator_status=[FAILED],
context_status=[], timestamp=[1460568243519], recoveryFlag =[False].


On Wed, Apr 13, 2016 at 11:42 AM, Markus Weimer <notifications@github.com>
wrote:

> @dkm2110 <https://github.com/dkm2110> I uploaded it to OneDrive
> <https://onedrive.live.com/redir?resid=5801726772BFC3DA!273888&authkey=!ACMrVHlIHAHvCi8&ithint=file%2czip>.
> Just be aware that it is a 160MB file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/940#issuecomment-209589249>
>

markusweimer commented:
Let's file a JIRA for this and merge this one. Agreed?

dkm2110 commented:
Yupp.

On Wed, Apr 13, 2016 at 11:50 AM, Markus Weimer <notifications@github.com>
wrote:

> Let's file a JIRA for this and merge this one. Agreed?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/940#issuecomment-209592420>
>

tcNickolas commented:
+1

markusweimer commented:
Done on both accounts, the JIRA is [REEF-1338](https://issues.apache.org/jira/browse/REEF-1338)
